docs/source/notes/randomness.rst

```diff
@@ -2,75 +2,115 @@ Reproducibility
-nondeterministically on the CUDA backend because :meth:`repeat_interleave`
-is implemented using :meth:`index_select`, the backward path for
-which is implemented using :meth:`index_add_`, which is known to operate
-nondeterministically (in the forward direction) on the CUDA backend (see above).
-
-cuDNN
-.....
-When running on the cuDNN backend, two further options must be set::
-
-    torch.backends.cudnn.deterministic = True
-    torch.backends.cudnn.benchmark = False
+    import numpy as np
+    np.random.seed(0)
 
-On some versions of cuDNN and CUDA, RNN and LSTM may have non-deterministic behavior.
+However, some applications and libraries may use NumPy Random Generator objects,
+not the global RNG
+(`<https://numpy.org/doc/stable/reference/random/generator.html>`_), and those will
+need to be seeded consistently as well.
+
+If you are using any other libraries that use random number generators, refer to
+the documentation for those libraries to see how to set consistent seeds for them.
+
+CUDA convolution benchmarking
+-----------------------------
+CUDA convolution operations have a performance enhancement feature that can introduce
+randomness between multiple executions of an application. You can disable this feature
+with :code:`torch.backends.cudnn.benchmark = False`.
+
+However, if you do not need reproducibility across multiple executions of your application,
+then it would be beneficial to enable this benchmarking feature.
+
+Any time a CUDA convolution operation is called with a new set of size parameters and
+:code:`torch.backends.cudnn.benchmark = True` is set, CUDA will run multiple different
+implementations of the operation and benchmark them. The fastest implementation for your
+patform will then be cached and used for all further CUDA convolution operations that
+use that same set of size parameters, until the PyTorch program quits. Since the runtime
+of a convolution can vary, the fastest implementation could change each time the
+benchmark is run.
+
+Note that this setting is different from the :code:`torch.backends.cudnn.deterministic`
+setting discussed below.
+
+Avoiding nondeterministic algorithms
+....................................
+:meth:`torch.set_deterministic` lets you configure PyTorch to use deterministic
+algorithms instead of nondeterministic ones where available, and to throw an error
+if an operation is known to be nondeterministic (and without a deterministic
+alternative).
+
+Please check the documentation for :meth:`torch.set_deterministic()` for a full
+list of affected operations. If an operation does not act correctly according to
+the documentation, or if you need a deterministic implementation of an operation
+that does not have one, please submit an issue:
+`<https://github.com/pytorch/pytorch/issues?q=label:%22topic:%20determinism%22>`_
+
+For example, running the nondeterministic CUDA implementation of :meth:`index_add_`
+will throw an error::
+
+    >>> import torch
```

Comment:
Is there also a positive example where a deterministic algorithm is used instead?